0.6.5 Safer rsGxsChannel API

G10h4ck opened the pull request:
Protect data members with mutexes
Print messages when errors happens

G10h4ck committed:
Safer rsGxsChannel API

Protect data members with mutexes
Print messages when errors happens

G10h4ck committed:
Safer RsGxsChannel API

Deprecated old method which exposed interna async mechanism to the API
  users, making their and out life difficult
Things that really need to be async like turtle search/requests now accept
  callbacks, so the caller can be notified everytime some result is got
  back
Implement RsThread::async commodity wrapper to execute blocking API
  calls without blocking the caller, this could be optimized
  trasparently using a thread pool if necessary
Added hints into some retroshare-gui files on how to use RsThread::async
  thoghether with QMetaObject::invokeMethod and blocking RetroShare API
  to simplyfy interaction between GUI and libretroshare

G10h4ck committed:
Fix compilation with Qt older then 5.10

G10h4ck committed:
Fix compilation with Qt older then 5.7

G10h4ck committed:
Fix typo in documentation

G10h4ck committed:
gxschannels better debugging message

PhenomRetroShare commented on libretroshare/src/retroshare/rsgxschannels.h:
```diff
@@ -138,180 +199,186 @@ class RsGxsChannels: public RsGxsIfaceHelper, public RsGxsCommentService
 	 * @param[in] peers peers to which share the key
 	 * @return false on error, true otherwise
 	 */
+	RS_DEPRECATED_FOR(shareChannelKeys)
```
When it is deprecated, is it possible to say how to do now to replace it?

G10h4ck replied:
the `_FOR(shareChannelKeys)` is a suggestion on how to replace it ;)

PhenomRetroShare commented on libretroshare/src/retroshare/rsgxschannels.h:
```diff
@@ -138,180 +199,186 @@ class RsGxsChannels: public RsGxsIfaceHelper, public RsGxsCommentService
 	 * @param[in] post
 	 * @return false on error, true otherwise
 	 */
+	RS_DEPRECATED
```
No _FOR here?

G10h4ck replied:
because the right version to use has the same name

PhenomRetroShare replied:
But, is it possible to get class handler too?
I see that for RsGxsIfaceHelper::getTokenService() which is deprecated for RsGenExchange::getTokenService(), but not easy to know...
One more thing, in /libretroshare/src/util/rsdeprecate.h, you don't take care of clang. It's why I don't have message FOR. ;)

G10h4ck replied:
What is inside the parentheses of `RS_DEPRECATED_FOR()` get directly passed trough per-processor stringizer operator `#` and then printed, so you probably can put almost anything within those parentheses (test it please), about taking care of clang I haven't a comfortable clang based tool-chain to test it so please make a PR to take care of it ;)

PhenomRetroShare replied:
Where did you found __GNUC__ value starting using __deprecated__(message)?
Because here, with Clang v7.0.0, __GNUC__ == 4 && __GNUC_MINOR__ == 2 and message works:
/libretroshare/src/services/p3gxscircles.cc:1636: warning: 'getTokenService' is deprecated: Use 'RsGenExchange::getTokenService()' instead [-Wdeprecated-declarations]
GCC can handle it since 4.9:
https://gcc.gnu.org/gcc-4.9/changes.html#cxx

G10h4ck replied:
Don't remember at all, anyway take in account that `GNUC`, ` GNUC_MINOR` ans similar defined by clang are mostly meaningless, clang has it's own way to check for version, and probably define those macro too for some kind of retro-compatibility, but we must not rely on those when compiling with clang.

G10h4ck replied:
also the page you linked talks about  C++1y deprecated attribute which most probably is not supported when compiling in C++11 mode (which is how we compile)

G10h4ck committed:
RsGxsDataAccess filtered getters do not return all messages if none matches

This was causing scary error message in RsGxsChannels with autodoanload
  enabled, when all messages where already processed none matched the
  filter so all messages where returned,
  making p3GxsChannels::handleUnprocessedPost furious
  more details into this forum post
  retroshare://forum?name=Scary%20message%2C%20but%20which%20doesn%27t%20seem%20to%20be%20the%20source%20of%20the%20problems%20Was%3A%20More%20GXS%20strange%20messages&id=8fd22bd8f99754461e7ba1ca8a727995&msgid=04f10ff97f761c6840c33f1610cb050f0f73da8d

G10h4ck committed:
Better include in RsGxsChannels

felisucoibi commented:
could be nice the same for forums.
